File: frontend/src/types/fhir.ts
export type AdministrativeGender = 'male' | 'female' | 'other' | 'unknown';
export type ContactPointSystem = 'phone' | 'fax' | 'email' | 'pager' | 'url' | 'sms' | 'other';
export type ContactPointUse = 'home' | 'work' | 'temp' | 'old' | 'mobile';
export type AddressUse = 'home' | 'work' | 'temp' | 'old' | 'billing';
export type AddressType = 'both' | 'physical' | 'postal';
export type BundleType = 'document' | 'message' | 'transaction' | 'transaction-response' | 'batch' | 'batch-response' | 'history' | 'searchset' | 'collection';
export type ObservationStatus = 'registered' | 'preliminary' | 'final' | 'amended' | 'corrected' | 'cancelled' | 'entered-in-error' | 'unknown';

export interface Coding {
  system: string;
  code: string;
  display?: string;
  userSelected?: boolean;
}

export interface CodeableConcept {
  coding?: Coding[];
  text?: string;
}

export interface Quantity {
  value: number;
  comparator?: '<' | '<=' | '>=' | '>';
  unit?: string;
  system?: string;
  code?: string;
}

export interface Reference {
  reference: string;
  type?: 'Patient' | 'Observation' | 'Condition' | 'Bundle' | 'Practitioner';
  display?: string;
}

export interface Identifier {
  use?: 'usual' | 'official' | 'temp' | 'secondary' | 'old';
  type?: CodeableConcept;
  system: string;
  value: string;
  period?: Period;
  assigner?: Reference;
}

export interface HumanName {
  use?: 'usual' | 'official' | 'temp' | 'nickname' | 'anonymous' | 'old' | 'maiden';
  text?: string;
  family?: string;
  given?: string[];
  prefix?: string[];
  suffix?: string[];
  period?: Period;
}

export interface ContactPoint {
  system: ContactPointSystem;
  value: string;
  use?: ContactPointUse;
  rank?: number;
  period?: Period;
}

export interface Address {
  use?: AddressUse;
  type?: AddressType;
  text?: string;
  line: string[];
  city?: string;
  district?: string;
  state?: string;
  postalCode?: string;
  country?: string;
  period?: Period;
}

export interface Period {
  start?: string;
  end?: string;
}

export interface Contact {
  relationship: CodeableConcept[];
  name: HumanName;
  telecom: ContactPoint[];
  address: Address[];
  gender?: AdministrativeGender;
  organization?: Reference;
  period?: Period;
}

export interface Communication {
  language: CodeableConcept;
  preferred?: boolean;
}

export interface Extension {
  url: string;
  valueString?: string;
  valueBoolean?: boolean;
  valueCode?: string;
  valueUri?: string;
  valueDate?: string;
  valueDateTime?: string;
  valueInteger?: number;
  valueDecimal?: number;
  valueQuantity?: Quantity;
  extension?: Extension[];
}

export interface Meta {
  versionId?: string;
  lastUpdated?: string;
  profile?: string[];
  tag?: Coding[];
}

export interface PatientResource {
  resourceType: 'Patient';
  id: string;
  meta?: Meta;
  identifier: Identifier[];
  active?: boolean;
  name: HumanName[];
  telecom: ContactPoint[];
  gender: AdministrativeGender;
  birthDate?: string;
  deceasedBoolean?: boolean;
  deceasedDateTime?: string;
  address: Address[];
  maritalStatus?: CodeableConcept;
  multipleBirthBoolean?: boolean;
  contact: Contact[];
  communication: Communication[];
  extension?: Extension[];
}

export interface ObservationComponent {
  code: CodeableConcept;
  valueQuantity?: Quantity;
  valueString?: string;
  valueCodeableConcept?: CodeableConcept;
}

export interface ObservationResource {
  resourceType: 'Observation';
  id: string;
  status: ObservationStatus;
  category: CodeableConcept[];
  code: CodeableConcept;
  subject: Reference;
  effectiveDateTime: string;
  valueQuantity?: Quantity;
  valueString?: string;
  component: ObservationComponent[];
  extension?: Extension[];
}

export interface BundleEntry {
  fullUrl?: string;
  resource: PatientResource | ObservationResource;
  request?: {
    method: 'GET' | 'POST' | 'PUT' | 'PATCH' | 'DELETE';
    url: string;
  };
}

export interface BundleResource {
  resourceType: 'Bundle';
  id: string;
  type: BundleType;
  timestamp?: string;
  total?: number;
  entry: BundleEntry[];
}

export interface ValidationError {
  severity: 'error' | 'warning' | 'information';
  location: string;
  message: string;
  element?: string;
  fhirUri?: string;
}

export interface SuggestionAction {
  type: 'add-identifier' | 'add-observation' | 'set-field';
  target?: string;
  value?: unknown;
  identifier?: Identifier;
  observation?: ObservationResource;
}

export interface Suggestion {
  field: string;
  suggestion: string;
  confidence: number;
  code_example?: unknown;
  action?: SuggestionAction;
}

export interface OpenMRSMapping {
  uuid: string;
  identifiers: Array<{ system: string; value: string; type?: string }>;
  names: Array<{ given: string[]; family?: string; use?: string }>;
  gender: AdministrativeGender;
  birthdate?: string;
  address?: string;
}

export const FHIR_SYSTEMS = {
  identifierType: 'http://terminology.hl7.org/CodeSystem/v2-0203',
  observationCategory: 'http://terminology.hl7.org/CodeSystem/observation-category',
  loinc: 'http://loinc.org',
  snomed: 'http://snomed.info/sct',
  ucum: 'http://unitsofmeasure.org',
  administrativeGender: 'http://hl7.org/fhir/ValueSet/administrative-gender',
  abhaIdentifier: 'http://abdm.gov.in/fhir/NamingSystem/abha',
  hospitalIdentifier: 'http://hospital.example.org/mrn'
} as const;

export function pathToDomId(path: string): string {
  return path.replace(/[^a-zA-Z0-9]+/g, '-').replace(/^-+|-+$/g, '').toLowerCase();
}

export function makeCoding(system: string, code: string, display?: string): Coding {
  return { system, code, display };
}

export function makeCodeableConcept(coding: Coding[], text?: string): CodeableConcept {
  return { coding, text };
}

export function createPatientResource(id = 'pat-001'): PatientResource {
  return {
    resourceType: 'Patient',
    id,
    meta: {
      profile: ['http://hl7.org/fhir/StructureDefinition/Patient']
    },
    identifier: [
      {
        use: 'official',
        type: makeCodeableConcept([
          makeCoding(FHIR_SYSTEMS.identifierType, 'MR', 'Medical record number')
        ], 'Medical Record Number'),
        system: FHIR_SYSTEMS.hospitalIdentifier,
        value: '12345'
      }
    ],
    active: true,
    name: [
      {
        use: 'official',
        family: 'Devi',
        given: ['Asha'],
        prefix: ['Ms.']
      }
    ],
    telecom: [
      { system: 'phone', value: '+91-9876543210', use: 'mobile' }
    ],
    gender: 'female',
    birthDate: '2020-01-15',
    address: [
      {
        use: 'home',
        type: 'physical',
        line: ['Village Ramnagar'],
        city: 'Shimla',
        state: 'Himachal Pradesh',
        postalCode: '171001',
        country: 'IN'
      }
    ],
    maritalStatus: makeCodeableConcept([
      makeCoding(FHIR_SYSTEMS.snomed, '365581002', 'Married')
    ], 'Married'),
    multipleBirthBoolean: false,
    contact: [
      {
        relationship: [makeCodeableConcept([
          makeCoding(FHIR_SYSTEMS.snomed, '103495007', 'Emergency contact')
        ], 'Emergency contact')],
        name: {
          family: 'Kumar',
          given: ['Ravi']
        },
        telecom: [{ system: 'phone', value: '+91-9000000000', use: 'mobile' }],
        address: [{ line: ['Village Ramnagar'], city: 'Shimla', country: 'IN' }]
      }
    ],
    communication: [
      {
        language: makeCodeableConcept([
          makeCoding('urn:ietf:bcp:47', 'en', 'English')
        ], 'English'),
        preferred: true
      }
    ]
  };
}

export function createBloodPressureObservation(
  subjectReference: string,
  systolic = 120,
  diastolic = 80,
  effectiveDateTime = new Date().toISOString()
): ObservationResource {
  return {
    resourceType: 'Observation',
    id: `obs-${Math.random().toString(36).slice(2, 10)}`,
    status: 'final',
    category: [
      makeCodeableConcept([
        makeCoding(FHIR_SYSTEMS.observationCategory, 'vital-signs', 'Vital Signs')
      ], 'Vital Signs')
    ],
    code: makeCodeableConcept([
      makeCoding(FHIR_SYSTEMS.loinc, '85354-9', 'Blood pressure panel')
    ], 'Blood pressure panel'),
    subject: {
      reference: subjectReference,
      type: 'Patient'
    },
    effectiveDateTime,
    component: [
      {
        code: makeCodeableConcept([
          makeCoding(FHIR_SYSTEMS.loinc, '8480-6', 'Systolic blood pressure')
        ], 'Systolic blood pressure'),
        valueQuantity: {
          value: systolic,
          unit: 'mmHg',
          system: FHIR_SYSTEMS.ucum,
          code: 'mm[Hg]'
        }
      },
      {
        code: makeCodeableConcept([
          makeCoding(FHIR_SYSTEMS.loinc, '8462-4', 'Diastolic blood pressure')
        ], 'Diastolic blood pressure'),
        valueQuantity: {
          value: diastolic,
          unit: 'mmHg',
          system: FHIR_SYSTEMS.ucum,
          code: 'mm[Hg]'
        }
      }
    ]
  };
}

export function createCustomObservation(
  subjectReference: string,
  coding: Coding,
  value: number,
  unit: string,
  effectiveDateTime = new Date().toISOString()
): ObservationResource {
  return {
    resourceType: 'Observation',
    id: `obs-${Math.random().toString(36).slice(2, 10)}`,
    status: 'final',
    category: [
      makeCodeableConcept([
        makeCoding(FHIR_SYSTEMS.observationCategory, 'vital-signs', 'Vital Signs')
      ], 'Vital Signs')
    ],
    code: makeCodeableConcept([coding], coding.display),
    subject: {
      reference: subjectReference,
      type: 'Patient'
    },
    effectiveDateTime,
    valueQuantity: {
      value,
      unit,
      system: FHIR_SYSTEMS.ucum,
      code: unit
    },
    component: []
  };
}

export function createBundleResource(
  patient: PatientResource,
  observations: ObservationResource[],
  type: BundleType = 'collection'
): BundleResource {
  const entries: BundleEntry[] = [
    {
      fullUrl: `urn:uuid:${patient.id}`,
      resource: patient
    },
    ...observations.map(observation => ({
      fullUrl: `urn:uuid:${observation.id}`,
      resource: observation
    }))
  ];

  return {
    resourceType: 'Bundle',
    id: `bundle-${Math.random().toString(36).slice(2, 10)}`,
    type,
    timestamp: new Date().toISOString(),
    total: entries.length,
    entry: entries
  };
}

export function createOpenMRSMapping(patient: PatientResource): OpenMRSMapping {
  const displayName = patient.name[0];
  return {
    uuid: patient.id,
    identifiers: patient.identifier.map(identifier => ({
      system: identifier.system,
      value: identifier.value,
      type: identifier.type?.text ?? identifier.type?.coding?.[0]?.code
    })),
    names: patient.name.map(name => ({
      given: name.given ?? [],
      family: name.family,
      use: name.use
    })),
    gender: patient.gender,
    birthdate: patient.birthDate,
    address: displayName ? patient.address[0]?.line?.join(', ') : undefined
  };
}

File: app/schemas/fhir.py
"""FHIR resource and response schemas used by the backend API."""

from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field


class FHIRBaseModel(BaseModel):
    model_config = ConfigDict(extra="ignore")


class Coding(FHIRBaseModel):
    system: str
    code: str
    display: str | None = None
    userSelected: bool | None = None


class CodeableConcept(FHIRBaseModel):
    coding: list[Coding] = Field(default_factory=list)
    text: str | None = None


class Quantity(FHIRBaseModel):
    value: float
    comparator: Literal["<", "<=", ">=", ">"] | None = None
    unit: str | None = None
    system: str | None = None
    code: str | None = None


class Reference(FHIRBaseModel):
    reference: str
    type: str | None = None
    display: str | None = None


class Period(FHIRBaseModel):
    start: str | None = None
    end: str | None = None


class Identifier(FHIRBaseModel):
    use: Literal["usual", "official", "temp", "secondary", "old"] | None = None
    type: CodeableConcept | None = None
    system: str
    value: str
    period: Period | None = None
    assigner: Reference | None = None


class HumanName(FHIRBaseModel):
    use: Literal["usual", "official", "temp", "nickname", "anonymous", "old", "maiden"] | None = None
    text: str | None = None
    family: str | None = None
    given: list[str] = Field(default_factory=list)
    prefix: list[str] = Field(default_factory=list)
    suffix: list[str] = Field(default_factory=list)
    period: Period | None = None


class ContactPoint(FHIRBaseModel):
    system: Literal["phone", "fax", "email", "pager", "url", "sms", "other"]
    value: str
    use: Literal["home", "work", "temp", "old", "mobile"] | None = None
    rank: int | None = None
    period: Period | None = None


class Address(FHIRBaseModel):
    use: Literal["home", "work", "temp", "old", "billing"] | None = None
    type: Literal["both", "physical", "postal"] | None = None
    text: str | None = None
    line: list[str] = Field(default_factory=list)
    city: str | None = None
    district: str | None = None
    state: str | None = None
    postalCode: str | None = None
    country: str | None = None
    period: Period | None = None


class Contact(FHIRBaseModel):
    relationship: list[CodeableConcept] = Field(default_factory=list)
    name: HumanName
    telecom: list[ContactPoint] = Field(default_factory=list)
    address: list[Address] = Field(default_factory=list)
    gender: Literal["male", "female", "other", "unknown"] | None = None
    organization: Reference | None = None
    period: Period | None = None


class Communication(FHIRBaseModel):
    language: CodeableConcept
    preferred: bool | None = None


class Meta(FHIRBaseModel):
    versionId: str | None = None
    lastUpdated: str | None = None
    profile: list[str] = Field(default_factory=list)
    tag: list[Coding] = Field(default_factory=list)


class Extension(FHIRBaseModel):
    url: str
    valueString: str | None = None
    valueBoolean: bool | None = None
    valueCode: str | None = None
    valueUri: str | None = None
    valueDate: str | None = None
    valueDateTime: str | None = None
    valueInteger: int | None = None
    valueDecimal: float | None = None
    valueQuantity: Quantity | None = None
    extension: list["Extension"] = Field(default_factory=list)


class PatientResource(FHIRBaseModel):
    resourceType: Literal["Patient"] = "Patient"
    id: str
    meta: Meta | None = None
    identifier: list[Identifier] = Field(default_factory=list)
    active: bool | None = None
    name: list[HumanName] = Field(default_factory=list)
    telecom: list[ContactPoint] = Field(default_factory=list)
    gender: Literal["male", "female", "other", "unknown"]
    birthDate: str | None = None
    deceasedBoolean: bool | None = None
    deceasedDateTime: str | None = None
    address: list[Address] = Field(default_factory=list)
    maritalStatus: CodeableConcept | None = None
    multipleBirthBoolean: bool | None = None
    contact: list[Contact] = Field(default_factory=list)
    communication: list[Communication] = Field(default_factory=list)
    extension: list[Extension] = Field(default_factory=list)


class ObservationComponent(FHIRBaseModel):
    code: CodeableConcept
    valueQuantity: Quantity | None = None
    valueString: str | None = None
    valueCodeableConcept: CodeableConcept | None = None


class ObservationResource(FHIRBaseModel):
    resourceType: Literal["Observation"] = "Observation"
    id: str
    status: Literal["registered", "preliminary", "final", "amended", "corrected", "cancelled", "entered-in-error", "unknown"]
    category: list[CodeableConcept] = Field(default_factory=list)
    code: CodeableConcept
    subject: Reference
    effectiveDateTime: str
    valueQuantity: Quantity | None = None
    valueString: str | None = None
    component: list[ObservationComponent] = Field(default_factory=list)
    extension: list[Extension] = Field(default_factory=list)


class BundleEntryRequest(FHIRBaseModel):
    method: Literal["GET", "POST", "PUT", "PATCH", "DELETE"]
    url: str


class BundleEntry(FHIRBaseModel):
    fullUrl: str | None = None
    resource: PatientResource | ObservationResource
    request: BundleEntryRequest | None = None


class BundleResource(FHIRBaseModel):
    resourceType: Literal["Bundle"] = "Bundle"
    id: str
    type: Literal[
        "document",
        "message",
        "transaction",
        "transaction-response",
        "batch",
        "batch-response",
        "history",
        "searchset",
        "collection",
    ]
    timestamp: str | None = None
    total: int | None = None
    entry: list[BundleEntry] = Field(default_factory=list)


class ValidationIssue(FHIRBaseModel):
    severity: Literal["error", "warning", "information"]
    location: str
    message: str
    element: str | None = None
    fhirUri: str | None = None


class SuggestionAction(FHIRBaseModel):
    type: Literal["add-identifier", "add-observation", "set-field"]
    target: str | None = None
    value: Any | None = None
    identifier: Identifier | None = None
    observation: ObservationResource | None = None


class SuggestionItem(FHIRBaseModel):
    field: str
    suggestion: str
    confidence: float = Field(ge=0.0, le=1.0)
    code_example: Any | None = None
    action: SuggestionAction | None = None
    original: Any | None = None
    suggested: Any | None = None
    reason: str | None = None
    needs_review: bool = False


class OpenMRSMapping(FHIRBaseModel):
    uuid: str
    identifiers: list[dict[str, Any]] = Field(default_factory=list)
    names: list[dict[str, Any]] = Field(default_factory=list)
    gender: Literal["male", "female", "other", "unknown"]
    birthdate: str | None = None
    address: str | None = None


class LegacyPatientRecord(FHIRBaseModel):
    id: str
    name: str
    age: int
    gender: Literal["male", "female", "other", "unknown"]
    vitals: dict[str, Any] = Field(default_factory=dict)
    diagnoses: list[str] = Field(default_factory=list)
    phone: str | None = None
    pregnant: bool | None = None


class FHIRValidationRequest(FHIRBaseModel):
    patient: PatientResource
    observations: list[ObservationResource] = Field(default_factory=list)
    bundle: BundleResource | None = None
    profile: str = "Base FHIR R4 Patient"


class ValidationResponse(FHIRBaseModel):
    quality_score: int = Field(ge=0, le=100)
    issues: list[ValidationIssue] = Field(default_factory=list)
    fhir_compliant: bool
    profile: str = "Base FHIR R4 Patient"
    patient: PatientResource | None = None
    observations: list[ObservationResource] = Field(default_factory=list)
    bundle: BundleResource | None = None
    openmrs_mapping: OpenMRSMapping | None = None
    suggestions: list[SuggestionItem] = Field(default_factory=list)


class FHIRCheckRequest(FHIRBaseModel):
    patient: PatientResource | None = None
    observations: list[ObservationResource] = Field(default_factory=list)
    bundle: BundleResource | None = None
    profile: str = "Base FHIR R4 Patient"
    record: dict[str, Any] = Field(default_factory=dict)
